src/libusermode/running.cpp

```diff
@@ -0,0 +1,490 @@
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include <fstream>
+#include <sstream>
+#include <map>
+#include <string>
+#include <optional>
+
+#include <config.h>
+#include <glib.h>
+#include <inttypes.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <libdrakvuf/libdrakvuf.h>
+#include <assert.h>
+
+#include "userhook.hpp"
+#include "uh-private.hpp"
+
+
+struct er_data_t
+{
+    // Arguments provided by the user.
+    addr_t target_process;
+    std::string dll_name;
+    std::string func_name;
+    callback_t cb;
+    void* extra;
+
+    // Additional data. Stored here for optimalization.
+    target_hook_state state;
+    vmi_pid_t target_process_pid;
+    addr_t target_process_dtb;
+    addr_t func_addr;
+
+    // We need to pass this around as we need offsets.
+    userhook* userhook_plugin;
+
+    er_data_t(userhook* userhook_plugin, addr_t target_process, vmi_pid_t target_process_pid,
+              std::string dll_name, std::string func_name, callback_t cb, void* extra):
+        target_process(target_process), dll_name(dll_name), func_name(func_name), cb(cb),
+        extra(extra), state(HOOK_FIRST_TRY), target_process_pid(target_process_pid),
+        userhook_plugin(userhook_plugin) {}
+};
+
+
+static
+void free_trap(drakvuf_trap_t* trap)
+{
+    if (!trap)
+        return;
+
+    if (trap->data)
+        delete (er_data_t*) trap->data;
+
+    delete trap;
+}
+
+
+/**
+ * Searches process's InLoadOrderModuleList for given dll library and
+ * sets res_dll_base to its base address. Returns true on success.
+ */
+static
+bool get_dll_base(
+    drakvuf_t drakvuf,
+    const size_t* offsets,
+    addr_t process_base,
+    std::string dll_name,
+    addr_t* res_dll_base)
+{
+    addr_t process_dtb = 0;
+    if (!drakvuf_get_process_dtb(drakvuf, process_base, &process_dtb))
+        return false;
+
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = process_dtb
+    };
+    vmi_lock_guard lg(drakvuf);
```

Comment:
Please add another scope for `lock_guard`. It explicitly shows where we need the lock (even if it is till end of the function).